### PR TITLE
docs: add concepts.md and internal.md

### PR DESCRIPTION
**Pull request by frsyuki**
Concepts and internal architecture documents for Digdag administrators and developers.

### CODE REVIEW

**Comment by frsyuki** [comment]
@danielnorberg  @komamitsu could you let me know what should be added to the docs?

File: digdag-docs/src/concepts.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+
+## Tasks
+
+When an attempt of a session starts, a workflow is transformed into a set of tasks. A task has dependency; task +dump depends on +process1 and +process2, task +process1 and +process2 depend on +prepare, etc. Digdag understands the dependencies and run the tasks in order.
+
+
+## Export and store parameters
+
+There are 3 kinds of parameters for a task: local parameters directly set to the task, parameters exported from parent tasks, and parameters stored by previous tasks. They are merged into one object when a task runs.
```
Comment (komamitsu) [inline digdag-docs/src/concepts.md]: How about clarifying the priority of these 3 operators?

File: digdag-docs/src/internal.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+This guide explains implementation details of Digdag. Understanding the internal architecture is helpful for Digdag system administrators and developers to understand why Digdag behaves as described at [architecture](architecture.html) page.
+
+## Task logging
+
+When a task runs, a logger (SLF4j + logback) collects its log messages and passes them to a task logger set at a thread-local storage. When a task starts, task logger creates a new file and writes messages to the file. As the file grows larger than certain limit, it uploads the file to a storage (note: there is an optimization for local execution with local file system logger: logs are directly appended to the final destionation file). LogServer is the SPI interface that implements this storage.
+
+When a client uploads a file, the client first requests a direct upload URL. If LogServer supports a temporary pre-signed HTTP URL to upload files, server returns the URL. Then, the client uploads the file to the URL directly. Otherwise, the client uploads the file to a digdag server, and the digdag server uploads the contents using LogServer. LogServer stores the file using full name of the task as its path name.
```
Comment (komamitsu) [inline digdag-docs/src/internal.md]: Maybe we'd better clarify what's `client` in this context.

File: digdag-docs/src/internal.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+`tasks`, `task_details`, and `task_dependencies` tables stores tasks of running attempts. `tasks` table stores state of tasks. `task_details` stores config and parameters of tasks, and `task_dependencies` stores dependency between multiple tasks under an parent tasks. Workflow executor checks these tables periodically to run workflows. When an attempt finishes, its tasks will be removed from the tables and archived in `task_archives` table.
+
+
+## Task queue
+
+Task queue makes sure that a task runs at least once.
+
+When an attempt starts, WorkflowExecutor pushes the root task to a task queue using TaskQueueServer interface. When the task finished, WorkflowExecutor will get a callback when a task finishes at least once. As tasks become ready to execute, WorkflowExecutor pushes the tasks to a task queue.
+
+An agent uses TaskQueueClient interface to fetch tasks. The default implementation of TaskQueueClient is TaskQueueServer itself which fetches tasks from the underlaying storage directly. An intended implementation is HTTP-based client that fetches tasks through Digdag's REST API.
```
Comment (komamitsu) [inline digdag-docs/src/internal.md]: Should we explain `agent` before this chapter?

**Comment by komamitsu** [comment]
Adding diagrams to the doc would be great.

File: digdag-docs/src/concepts.md (hunk 1 above)
Comment (frsyuki) [inline digdag-docs/src/concepts.md]: 👍  added

File: digdag-docs/src/internal.md (hunk 2 above)
Comment (frsyuki) [inline digdag-docs/src/internal.md]: 👍  changed client to `the task logger`.

File: digdag-docs/src/internal.md (hunk 3 above)
Comment (frsyuki) [inline digdag-docs/src/internal.md]: 👍  added another chapter before this.

**Comment by frsyuki** [comment]
@komamitsu let me make it TODO for now...it will be presentation-driven.

**Pull request merged**